Running relayer: documentation (#2829)
paritytech/parity-bridges-common: docs/running-relayer.md

means that all supplied data in transaction is new and yet unknown to the target chain.

We have implemented a relayer that is able to craft such transactions. The rest of document contains a detailed
information on how to deploy this software on your own node.

## Relayers Concurrency

As it has been said above, we are not compensating cost of transactions that are not **useful**. For
example, if message `100` has already been delivered from Kusama Bridge Hub to Polkadot Bridge Hub, then another
transaction that delivers the same message `100` won't be **useful**. Hence, no compensation to relayer that
has submitted that second transaction.

But what if there are several relayers running? They are noticing the same queued message `100` and
simultaneously submit identical message delivery transactions. You may expect that there'll be one lucky
relayer, whose transaction would win the "race" and which will receive the compensation and reward. And
there'll be several other relayers, losing some funds on their unuseful transactions.

But actually, we have a solution that invalidates transactions of "unlucky" relayers before they are

> included into the block. So at least you may sure that you won't waste your tokens on duplicate transactions.

Reviewer: ```suggestion
included into the block. So at least you may be sure that you won't waste your funds on duplicate transactions.
```

`V3 { id: Concrete(0, Here), Fungible: <your-amount> }` to beneficiary `V3(0, X1(AccountId32(<your-account>)))`
on destination `V3(0, X1(Parachain(1002)))`. To estimate amounts you need, please refer to the [Costs](#costs)
section of the document.

## Registering your Relayer Account (Optional, But Please Read)

Bridge transactions are quite heavy and expensive. We want to minimize block space that can be occupied by
invalid bridge transactions and prioritize valid transactions over invalid. That is achieved by **optional**
relayer registration. Transactions, signed by relayers with active registration, gain huge priority boost.
In exchange, such relayers may be slashed if they submit **invalid** or **non-minimal** transaction.

Transactions, signed by relayers **without** active registration, on the other hand, receive no priority
boost. It means that if there is active registered relayer, most likely all transactions from unregistered
will be counted as **unuseful**, not included into the block and unregistered relayer won't get any reward
for his operations.

Before registering, you should know several things about your funds:

> - to register, you need to hold significant amount of tokens on your relayer account. As of now, it is

Reviewer: ```suggestion
- to register, you need to hold significant amount of funds on your relayer account. As of now, it is
```

| 700               | Inactive           | Your message delivery transactions are not boosted     |
| 701               | Inactive           | Your message delivery transactions are not boosted     |
| ...               | Inactive           | Your message delivery transactions are not boosted     |
| 1000              | Expired            | Your may submit a tx with the `deregister` call        |

</details>

So once you have enough funds on your account and have selected the `validTill` parameter value, you
could use the Polkadot JS apps to submit an extrinsic. If you want priority boost for your transactions
on the Kusama Bridge Hub, open the
[Polkadot JS Apps](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fkusama-bridge-hub-rpc.polkadot.io#/extrinsics)
and submit the `register` extrinsic from the `bridgeRelayers` pallet:

![Register Extrinsic](./bridge-relayers-register.png)

To deregister, submit the simple `deregister` extrinsic when registration is expired:

![Deregister Extrinsic](./bridge-relayers-deregister.png)

At any time, you can prolong your registration by calling the `register` with the larger `validTill`.

## Costs

> Your relayer account (on both Bridge Hubs) must hold enough tokens to be able to pay costs of bridge

Reviewer: ```suggestion
Your relayer account (on both Bridge Hubs) must hold enough funds to be able to pay costs of bridge
```

# argument to the `docker` binary
docker run \
    -v $BRIDGE_HUB_ROCOCO_KEY_FILE:/bhr.key \
    -v $BRIDGE_HUB_WESTEND_KEY_FILE:/bhw.key \
    $DOCKER_IMAGE \
    relay-headers-and-messages bridge-hub-rococo-bridge-hub-westend \
    --rococo-host $ROCOCO_HOST \
    --rococo-port $ROCOCO_PORT \
    $ROCOCO_IS_SECURE \
    --rococo-version-mode Auto \
    --bridge-hub-rococo-host $BRIDGE_HUB_ROCOCO_HOST \
    --bridge-hub-rococo-port $BRIDGE_HUB_ROCOCO_PORT \
    $BRIDGE_HUB_ROCOCO_IS_SECURE \
    --bridge-hub-rococo-version-mode Auto \
    --bridge-hub-rococo-signer-file /bhr.key \
    --bridge-hub-rococo-transactions-mortality 16 \
    --westend-host $WESTEND_HOST \
    --westend-port $WESTEND_PORT \
    $WESTEND_IS_SECURE \
    --westend-version-mode Auto \
    --bridge-hub-westend-host $BRIDGE_HUB_WESTEND_HOST \
    --bridge-hub-westend-port $BRIDGE_HUB_WESTEND_PORT \
    $BRIDGE_HUB_WESTEND_IS_SECURE \
    --bridge-hub-westend-version-mode Auto \
    --bridge-hub-westend-signer-file /bhw.key \
    --bridge-hub-westend-transactions-mortality 16 \
    --lane 00000002
```

### Starting your Polkadot <> Kusama Relayer

> *Work in progress, get back soon*

Reviewer: ```suggestion
*Work in progress, coming soon*
```

Our relayer provides some Prometheus metrics that you may convert into some fancy Grafana dashboards
and alerts. By default, metrics are exposed at port `9616`. To expose endpoint to the localhost, change
the docker command by adding following two lines:

```sh
docker run \
    ..
    -p 127.0.0.1:9616:9616 \ # tell Docker to bind container port 9616 to host port 9616
                             # and listen for connections on the host' localhost interface
    ..
    $DOCKER_IMAGE \
    relay-headers-and-messages bridge-hub-rococo-bridge-hub-westend \
    --prometheus-host 0.0.0.0 \ # tell `substrate-relay` binary to accept Prometheus endpoint
                                # connections from everywhere
    ..
```

You can find more info on configuring Prometheus and Grafana in the
[Monitor your node](https://wiki.polkadot.network/docs/maintain-guides-how-to-monitor-your-node)
guide from Polkadot wiki.

We have our own set of Grafana dashboards and alerts. You may use them for inspiration.
Please find them in this folder:

- for Rococo <> Westend bridge: [rococo-westend](https://github.com/paritytech/parity-bridges-common/tree/master/deployments/bridges/rococo-westend).

> - for Polkadot <> Kusama bridge: *work in progress, get back soon*

Reviewer: ```suggestion
- for Polkadot <> Kusama bridge: *work in progress, coming soon*
```

because compensations are paid from the account, controlled by relay chain governance and it can have no tokens
to compensate your useful actions. We'll be working on a proper process to resupply it on-time, but we can't
provide any guarantee until that process is well established.

## A Brief Introduction into Relayers and our Compensations Scheme

Omitting details, relayer is an offchain process that is connected to both bridged chains. It looks at the
outbound bridge messages queue and submits message delivery transactions to the target chain. There's a lot
of details behind that simple phrase - you could find more info in the
[High-Level Bridge Overview](./high-level-overview.md) document.

Reward that is paid to relayer has two parts. The first part static and is controlled by the governance.
It is rather small initially - e.g. you need to deliver `10_000` Kusama -> Polkadot messages to gain single
KSM token.

The other reward part is dynamic. So to deliver an XCM message from one BridgeHub to another, we'll need to
deliver several "things" from one chain to another (omitting details: this "thing" is an XCM message). Those

> "things" are submitted as transactions and every transaction has its cost, which is:

Reviewer 2: Nit: To me it seems a bit hard to follow. What do you mean by `we'll need to
deliver several "things" from one chain to another` ?

Author: The idea was to avoid unrelated details in this document - i.e. no mentions of storage and finality proofs, delivery and confirmation transactions, ... For that there's another document. But I agree - "things" aren't much better. I've removed this "things" "thing" and only left transactions

Reviewer 2: Looks good ! The only thing left is that `deliver submit` seems a bit redundant.
